### PR TITLE
Filter material from

### PR DESCRIPTION
**Pull request by bam241**
# Description

This aims to implement `MaterialFromFilter` as described in #2719

I inspired myself from `CellFromFilter`. I as particule only stores history of cell they travelled through and not materials, I used visited cells to list materials seen previously any the materials.

The only cons is that distribcell might contains multiple materials so this detection would be wrong. As particule might have travelled through a distribcell but not seen the requested materials.

`in Cell.h line 290-293`
```
  //! \brief Material(s) within this cell.
  //!
  //! May be multiple materials for distribcell.
  vector<int32_t> material_;
```

If you have any recommandation on how to implement this better, I will be happy to comply. If there is none, I can throw an Error or a Warning, stating that some cell contains multiple material so it is not possible to accurately filter by `MaterialFrom`.

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

_This work is sponsored by [First Light Fusion](https://firstlightfusion.com)._

### CODE REVIEW

File: include/openmc/particle_data.h [hunk 1]
```diff
@@ -435,6 +435,7 @@ class ParticleData {
   int& material() { return material_; }
   const int& material() const { return material_; }
   int& material_last() { return material_last_; }
+  const int& const_material_last() const { return material_last_; }
```
Comment (pshriwise) [inline include/openmc/particle_data.h:438]:
```suggestion
  const int& material_last() const { return material_last_; }
```

File: include/openmc/tallies/filter_material.h [hunk 2]
```diff
@@ -46,7 +46,7 @@ class MaterialFilter : public Filter {
 
   void set_materials(gsl::span<const int32_t> materials);
 
-private:
+protected:
```
Comment (pshriwise) [inline include/openmc/tallies/filter_material.h:49]: Why was this change necessary?

File: tests/regression_tests/surface_tally/test.py [hunk 3]
```diff
@@ -106,19 +106,19 @@ def __init__(self, *args, **kwargs):
 
         # Create partial current tallies from water to fuel
         # Filters
-        cell_from_filter = openmc.CellFromFilter(water)
+        mat_from_filter = openmc.MaterialFromFilter(borated_water)
```
Comment (pshriwise) [inline tests/regression_tests/surface_tally/test.py:109]: The fact that the results didn't change here is good! We'll probably want to do this with two tallies though so we're still testing the `CellFromFilter` and that the surface currents are the same when using either a `CellFromFilter` or the `MaterialFromFilter`.

**Review by pshriwise — CHANGES_REQUESTED** [review]
Nice addition @bam241! Just a few small questions/comments from me.

File: tests/regression_tests/surface_tally/test.py (hunk 3 above)
Comment (bam241) [reply tests/regression_tests/surface_tally/test.py:109]: there were 2 usage of the `openmc.CellFromFilter(water)`, so I figured I could divert 1, 
I can also implement an additional test.

File: include/openmc/tallies/filter_material.h (hunk 2 above)
Comment (bam241) [reply include/openmc/tallies/filter_material.h:49]: I need to use both `materials_` and `map_` that were private member of the `FilterMaterial` class, those change match what was done in the `FilterCell` class.

Comment (pshriwise) [reply include/openmc/tallies/filter_material.h:49]: Ah right, needed so they can be accessed in the inherited class. Thanks!

File: tests/regression_tests/surface_tally/test.py (hunk 3 above)
Comment (pshriwise) [reply tests/regression_tests/surface_tally/test.py:109]: Hrmmm yeah that's probably okay. Ideally, we'd have dedicated test cases for the Cell/MaterialFromFilter's rather than wrap them into this surface tally test, but that shouldn't be on you to fix since it's a pre-existing issue. Would you be willing to make an issue so we can track that though? (in case one of us magically has time to address it in the future) Thanks!

Comment (bam241) [reply tests/regression_tests/surface_tally/test.py:109]: I'll do that !
thanks for the review

**Review by pshriwise — APPROVED** [review]
This looks good to me. Thanks @bam241 for the new filter!

**Review by paulromano — APPROVED** [review]
Nice work on this @bam241. Thanks for the addition!
